File: nlp.py
from random import sample
from re import findall


class Nlp100_01:
    def q00(self):
        s = "stressed"
        print(s[::-1])

    def q01(self):
        s = "パタトクカシー"
        print(s[0::2])

    def q02(self):
        s1 = "パトカー"
        s2 = "タクシー"
        print(''.join(sum(zip(s1, s2), ())))

    def q03(self):
        s = "Now I need a drink, alcoholic of course, after the heavy lectures involving quantum mechanics."
        s = list(map(len, findall(r"\w+", s)))
        print(s)

    def q04(self):
        s = "Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can."
        arr = [1, 5, 6, 7, 8, 9, 15, 16, 19]
        s = s.split()
        ans = {}
        for x in range(1, len(s) + 1):
            if x in arr:
                ans[x] = s[x - 1][:1]
            else:
                ans[x] = s[x - 1][:2]
        print(ans)

    def q05_word_bi_gram(self, s, n):
        s = s.split()
        return [s[x:x + 2] for x in range(len(s) - 1)]

    def q05_letter_bi_gram(self, s, n):
        return [s[x:x + 2] for x in range(len(s) - 1)]

    def q05(self):
        s = "I am an NLPer"
        print(self.q05_word_bi_gram(s, 2))
        print(self.q05_letter_bi_gram(s, 2))

    def q06_letter_bi_gram(self, s, n):
        return [s[x:x + 2] for x in range(len(s) - 1)]

    def q06(self):
        s1 = "paraparaparadise"
        s2 = "paragraph"
        X = set(self.q06_letter_bi_gram(s1, 2))
        Y = set(self.q06_letter_bi_gram(s2, 2))
        print(f"{X = }")
        print(f"{Y = }")
        print(f"{X & Y = }")
        print(f"{X & Y = }")
        print(f"{X - Y = }")
        print(f"{'se' in X = }")
        print(f"{'se' in Y = }")

    def q07_tmp_new_str(self, x, y, z):
        return f"{x}時の{y}は{z}"

    def q07(self):
        print(self.q07_tmp_new_str(12, "気温", 22.4))

    def q08_cipher(self, s):
        res = ""
        for x in s:
            if x.islower():
                res += chr(219 - ord(x))
            else:
                res += x
        return res

    def q08(self):
        print("暗号化する文章を入力してください")
        s = input("=>")
        c = self.q08_cipher(s)
        print(f"暗号化: {c}")
        print(f"復号化: {self.q08_cipher(c)}")

    def q09(self):
        print("文章を入力してください")
        s = input("=>")
        ans = ""
        for x in s.split():
            if len(x) <= 4:
                ans += x
            else:
                ans += x[0] + "".join(sample(x[1:-1], len(x) - 2)) + x[-1]
            ans += " "
        print(ans)


nlp = Nlp100_01()
nlp.q00()
nlp.q01()
nlp.q02()
nlp.q03()
nlp.q04()
nlp.q05()
nlp.q06()
nlp.q07()
nlp.q08()
nlp.q09()

File: nlp.rb
require 'set'


class Nlp100
  def q00()
    s = "stressed"
    puts s.reverse
  end

  def q01()
    s = "パタトクカシー"
    ans = ""
    for i in [1, 3, 5, 7].map {|n| n - 1}
      ans += s[i]
    end
    puts ans
  end

  def q02()
    s1 = "パトカー"
    s2 = "タクシー"
    ans = ""
    for i in (0...4)
      ans += s1[i] + s2[i]
    end
    puts ans
  end

  def q03()
    s = "Now I need a drink, alcoholic of course, after the heavy lectures involving quantum mechanics."
    s = s.split(" ")
    ans = []
    for x in s
      x = x.slice(/\w+/)
      ans.push(x.length)
    end
    print ans
    puts
  end

  def q04()
    s = "Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can."
    s = s.split(" ")
    ls = [1, 5, 6, 7, 8, 9, 15, 16, 19]
    ls.map! {|n| n - 1}
    ans = {}
    for i in (0..(s.length - 1))
      if ls.include?(i)
        ans[i + 1] = s[i][0]
      else
        ans[i + 1] = s[i][0, 2]
      end
    end
    puts ans
  end

  def q05_word_bi_gram(s, n)
    s = s.split(" ")
    res = []
    for x in (0...(s.length - 1))
      res.push(s.slice(x..(x + (n - 1))))
    end
    return res
  end

  def q05_letter_bi_gram(s, n)
    res = []
    for x in (0...(s.length - 1))
      res.push(s.slice(x..(x + (n - 1))))
    end
    return res
  end

  def q05()
    s = "I am an NLPer"
    print q05_word_bi_gram(s, 2)
    puts
    print q05_letter_bi_gram(s, 2)
    puts
  end

  def q06_letter_bi_gram(s, n)
    res = []
    for x in (0...(s.length - 1))
      res.push(s.slice(x..(x + (n - 1))))
    end
    return res
  end

  def q06()
    s1 = "paraparaparadise"
    s2 = "paragraph"
    x = q06_letter_bi_gram(s1, 2).to_set
    y = q06_letter_bi_gram(s2, 2).to_set
    puts "X = #{x}"
    puts "Y = #{y}"
    puts "X + Y = #{x + y}"
    puts "X & Y = #{x & y}"
    puts "X - Y = #{x - y}"
    puts "'se' in X = #{x.include?("se")}"
    puts "'se' in Y = #{y.include?("se")}"
  end

  def q07_tmp_new_str(x, y, z)
    return "#{x}時の#{y}は#{x}"
  end

  def q07()
    puts q07_tmp_new_str(12, "気温", 22.4)
  end

  def q08_cipher(s)
    res = ""
    for x in s.split("")
      if (x =~ /^[a-z]+$/) != nil
        res += (219 - x.ord).chr
      else
        res += x
      end
    end
    return res
  end

  def q08()
    puts "暗号化する文章を入力してください"
    s = gets
    c = q08_cipher(s)
    puts "暗号化: #{c}"
    puts "復号化: #{q08_cipher(c)}"
  end

  def q09()
    puts "文章を入力してください"
    s = gets
    ans = ""
    for x in s.split(" ")
      if x.length <= 4
        ans += x
      else
        ans += x[0] + x.slice(1, x.length - 2).split("").shuffle.join("") + x[-1]
      end
      ans += " "
    end
    puts ans
  end
end


nlp = Nlp100.new
nlp.q00
nlp.q01
nlp.q02
nlp.q03
nlp.q04
nlp.q05
nlp.q06
nlp.q07
nlp.q08
nlp.q09
